### PR TITLE
A method to check for page stability when shiny is not busy

### PR DESCRIPTION
**Pull request by vedhav**
Introduces a private method for `TealAppDriver` called `wait_for_page_stability` which blocks the R process by sleeping until the page is unchanged till the stability period.

This is used in two places:
1. After a click is performed.
2. Before the visibility is checked.

This should potentially fix the CI errors in tmg and tmc.

Error in tmg:
```r
  ══ Failed tests ════════════════════════════════════════════════════════════════
  ── Error ('test-shinytest2-tm_misssing_data.R:73:3'): e2e - tm_missing_data: Default settings and visibility of the summary graph ──
  Error in `app_wait_for_idle(self, private, duration = duration, timeout = timeout)`: An error occurred while waiting for Shiny to be stable
  Backtrace:
      ▆
   1. └─app_driver$click(selector = app_driver$active_module_element("iris-filter_na")) at test-shinytest2-tm_misssing_data.R:73:3
   2.   └─self$wait_for_idle()
   3.     └─shinytest2:::app_wait_for_idle(self, private, duration = duration, timeout = timeout)
   4.       └─shinytest2:::app_abort(self, private, "An error occurred while waiting for Shiny to be stable")
   5.         └─rlang::abort(..., app = self, call = call)
  ── Error ('test-shinytest2-tm_misssing_data.R:123:3'): e2e - tm_missing_data: Validate functionality and UI response for 'By Variable Levels' ──
  Error in `app_wait_for_idle(self, private, duration = duration, timeout = timeout)`: An error occurred while waiting for Shiny to be stable
  Backtrace:
      ▆
   1. └─app_driver$set_active_module_input("iris-summary_type", "By Variable Levels") at test-shinytest2-tm_misssing_data.R:123:3
   2.   └─self$wait_for_idle()
   3.     └─shinytest2:::app_wait_for_idle(self, private, duration = duration, timeout = timeout)
   4.       └─shinytest2:::app_abort(self, private, "An error occurred while waiting for Shiny to be stable")
   5.         └─rlang::abort(..., app = self, call = call)
  
  [ FAIL 2 | WARN 5 | SKIP 0 | PASS 464 ]
```

Error in tmc:
```r
  ══ Failed tests ════════════════════════════════════════════════════════════════
  ── Failure ('test-shinytest2-tm_t_pp_basic_info.R:78:5'): e2e - tm_t_pp_basic_info: Deselection of patient_id throws validation error and table is not visible. ──
  app_driver$is_visible(...) is not FALSE
  
  `actual`:   TRUE 
  `expected`: FALSE
  ── Failure ('test-shinytest2-tm_t_pp_basic_info.R:118:3'): e2e - tm_t_pp_basic_info: Deselection of cov_var throws validation error. ──
  app_driver$is_visible(...) is not FALSE
  
  `actual`:   TRUE 
  `expected`: FALSE
  ── Failure ('test-shinytest2-tm_t_pp_laboratory.R:129:3'): e2e - tm_t_pp_laboratory: Deselection of patient_id throws validation error. ──
  app_driver$is_visible(...) is not FALSE
  
  `actual`:   TRUE 
  `expected`: FALSE
  ── Failure ('test-shinytest2-tm_t_pp_laboratory.R:166:3'): e2e - tm_t_pp_laboratory: Deselection of paramcd throws validation error. ──
  app_driver$is_visible(...) is not FALSE
  
  `actual`:   TRUE 
  `expected`: FALSE
  ── Failure ('test-shinytest2-tm_t_pp_laboratory.R:205:5'): e2e - tm_t_pp_laboratory: Deselection of param throws validation error. ──
  app_driver$is_visible(...) is not FALSE
  
  `actual`:   TRUE 
  `expected`: FALSE
  ── Failure ('test-shinytest2-tm_t_pp_laboratory.R:243:3'): e2e - tm_t_pp_laboratory: Deselection of timepoints throws validation error. ──
  app_driver$is_visible(...) is not FALSE
  
  `actual`:   TRUE 
  `expected`: FALSE
  ── Failure ('test-shinytest2-tm_t_pp_laboratory.R:282:3'): e2e - tm_t_pp_laboratory: Deselection of avalu throws validation error. ──
  app_driver$is_visible(...) is not FALSE
  
  `actual`:   TRUE 
  `expected`: FALSE
  ── Failure ('test-shinytest2-tm_t_pp_laboratory.R:321:3'): e2e - tm_t_pp_laboratory: Deselection of aval_var throws validation error. ──
  app_driver$is_visible(...) is not FALSE
  
  `actual`:   TRUE 
  `expected`: FALSE
  ── Failure ('test-shinytest2-tm_t_pp_laboratory.R:358:3'): e2e - tm_t_pp_laboratory: Deselection of arind throws validation error. ──
  app_driver$is_visible(...) is not FALSE
  
  `actual`:   TRUE 
  `expected`: FALSE
  
  [ FAIL 9 | WARN 8 | SKIP 0 | PASS 1966 ]
```

### CODE REVIEW

File: R/TealAppDriver.R [hunk 1]
```diff
@@ -575,6 +578,31 @@ TealAppDriver <- R6::R6Class( # nolint: object_name.
...
+        if (!identical(current_element, element_content)) {
+          element_content <- current_inner_html
+          times_run <- 0
+        } else if (times_run < max_times_run) {
+          times_run <- times_run + 1
+        } else {
+          break
+        }
+        Sys.sleep(check_interval / 1000)
+      }
```
Comment (kartikeyakirar) [inline R/TealAppDriver.R:605]: Should we move this method to private 🤔 ?

```suggestion
      element_content <- self$get_html(element_selector)
      times_run <- 0
      max_times_run <- stability_period / check_interval

      while (TRUE) {
        current_element <- self$get_html(element_selector)
        if (!identical(current_element, element_content)) {
          element_content <- current_inner_html
          times_run <- 0
        } else if (times_run < max_times_run) {
          times_run <- times_run + 1
        } else {
          break
        }
        Sys.sleep(check_interval / 1000)
      }
```
How about using "repeat with break"? Here is an alternative time comparison logic. Feel free to ignore.
```suggestion
      element_content <- self$get_html(element_selector)
  end_time <- Sys.time() + (stability_period / 1000)
  
  repeat {
    Sys.sleep(check_interval / 1000)
    current_element <- self$get_html(element_selector)
    
    if (!identical(current_element, element_content)) {
      element_content <- current_element
      end_time <- Sys.time() + (stability_period / 1000)
    } else if (Sys.time() >= end_time) {
      break
    }
  }
```

Comment (vedhav) [reply R/TealAppDriver.R]: Sure, this one is easy to interpret as well. I did move it to private and renamed it to just handle `body` change.

File: R/TealAppDriver.R [hunk 2]
```diff
@@ -645,6 +647,28 @@ TealAppDriver <- R6::R6Class( # nolint: object_name.
       }
 
       NULL # If there are not any supported filters
+    },
+    #' @description
+    #' Check if the page is stable without any `DOM` updates in the body of the app.
+    #' This is achieved by blocing the R process by sleeping until the page is unchanged till the `stability_period`.
+    #' @param stability_period (`numeric(1)`) The time in milliseconds to wait till the page to be stable.
+    #' @param check_interval (`numeric(1)`) The time in milliseconds to check for changes in the page.
+    #' The stability check is reset when a change is detected in the page after sleeping for check_interval.
```
Comment (kartikeyakirar) [inline R/TealAppDriver.R:656]:
```suggestion
    # @description
    # Check if the page is stable without any `DOM` updates in the body of the app.
    # This is achieved by blocing the R process by sleeping until the page is unchanged till the `stability_period`.
    # @param stability_period (`numeric(1)`) The time in milliseconds to wait till the page to be stable.
    # @param check_interval (`numeric(1)`) The time in milliseconds to check for changes in the page.
    # The stability check is reset when a change is detected in the page after sleeping for check_interval.
```

**Review by kartikeyakirar — APPROVED** [review]
LGTM!! @vedhav minor comment with before merging into main

test with example.

```
pkgload::load_all()

data <- teal_data()
data <- within(data, {
  IRIS <- iris
  MTCARS <- mtcars
})

app_driver <- TealAppDriver$new(
  data = data,
  modules = modules(
    example_module(datanames = "IRIS"),
    example_module(datanames = "MTCARS")
  )
)

# View the app in one window
app_driver$view()

# Here this will wait for 2 seconds stability period, During this time manually make changes to the app before 2 seconds and see if the R session is being blocked. If it is blocked the implementation is proper.
app_driver$.__enclos_env__$private$wait_for_page_stability(stability_period = 20000)
```

File: R/TealAppDriver.R (hunk 2 above)
Comment (vedhav) [reply R/TealAppDriver.R:656]: Having roxygen2 comments here is not going to help with documentation as R6 class methods are not documented by roxygen2. But, it is better for readability in the code and all other methods use this. I see that we've missed it in a private method `get_active_filter_selection` so I'll add them there as well. Good catch.

Comment (vedhav) [reply R/TealAppDriver.R]: Good catch, I'll remove them

Comment (vedhav) [reply R/TealAppDriver.R]: I just learned that for private methods roxygen comments do not work.

Comment (vedhav) [reply R/TealAppDriver.R]: Interesting that our roxygen doc CI is unable to catch this

**Pull request merged**